compiler/src/org.graalvm.compiler.core/src/org/graalvm/compiler/core/gen/NodeLIRBuilder.java

```diff
@@ -171,6 +174,10 @@ protected DebugInfoBuilder createDebugInfoBuilder(StructuredGraph graph, NodeVal
     @Override
     public Value operand(Node node) {
         Value operand = getOperand(node);
+        if (node instanceof ValueNode && operand instanceof SharedMatchValue) {
```

Comment:
Why is this necessary?

Reply:
This is to handle the case when `ConstantNode` is matched but not swallowed by the matched rule.  This is an example on aarch64:
```
 @MatchRule("(Add=op (UnsignedRightShift=x src Constant=shiftAmt1) (LeftShift src Constant=shiftAmt2))")
 public ComplexMatchResult rotationConstant(ValueNode op, ValueNode x, ValueNode src, ConstantNode shiftAmt1, ConstantNode shiftAmt2) {
    assert src.getStackKind().isNumericInteger();
    assert shiftAmt1.getStackKind().getBitCount() == 32;
    assert shiftAmt2.getStackKind().getBitCount() == 32;

    int shift1 = shiftAmt1.asJavaConstant().asInt();
    int shift2 = shiftAmt2.asJavaConstant().asInt();
    if (op instanceof AddNode && (0 == shift1 || 0 == shift2)) {
        return null;
    }
    if ((0 == shift1 + shift2) || (src.getStackKind().getBitCount() == shift1 + shift2)) {
    return builder -> {
                Value a = operand(src);
                Value b = x instanceof LeftShiftNode ? operand(shiftAmt2) : operand(shiftAmt1);
                return getArithmeticLIRGenerator().emitBinary(LIRKind.combine(a, b), AArch64ArithmeticOp.ROR, false, a, b);
            };
    }
    return null;
  }
```

The `ConstantNode` `shiftAmt1` and `shiftAmt2` are all matched to the rule. However, it's still used as an operand of the final `ROR`. See:
```
Value b = x instanceof LeftShiftNode ? operand(shiftAmt2) : operand(shiftAmt1);
```
The compiler cannot know the real usages inside of the `evaluate` function of the `ComplexMatchValue` during matching.

Another way to fix is to make it wrong when `operand()` of a matched `ConstantNode` is called. For the above rule, it needs some more modification at aarch64 backend which I don't think is friendly if fixing it.

Reply:
I think if it wants them as a constant then it needs to construct them in that form instead of going through operand.  Why can't it call emitBinaryConst directly using the JavaConstant from the node?  You can't safely do potentially lazy evaluation using doRoot since it might be called in a non-dominating position.

Reply:
Yes, I greatly agree with you. I was thinking that it needs more effort to fix the rules, since this case also exists on AMD64. It seems easier and simpler than what I thought before.  I will fix it later. Thanks!